### ISSUE TITLE
vagrant or virtualbox does not appear to allow the synced folder for host

### ISSUE THREAD

**Issue by alesparza** [CIS548/vagrant-env#1]
See Piazza [@10](https://piazza.com/class/krtmu8tans12en?cid=10) for the student request.  Student is unable to clone the repository into the /vagrant directory due to a permission error.  Both Robert and I ran into the same issue.  I was able to clone it into the home workspace ~, so it is not a github or repository issue.  I am running Windows 10, not sure about the other two.

Here's the log of my fresh install

Dox@Lighthouse MINGW64 /d/Files/Omnia/595/VM
$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Box 'ubuntu/focal64' could not be found. Attempting to find and install...
    default: Box Provider: virtualbox
    default: Box Version: >= 0
==> default: Loading metadata for box 'ubuntu/focal64'
    default: URL: https://vagrantcloud.com/ubuntu/focal64
==> default: Adding box 'ubuntu/focal64' (v20210826.0.0) for provider: virtualbox
    default: Downloading: https://vagrantcloud.com/ubuntu/boxes/focal64/versions/20210826.0.0/providers/virtualbox.box
==> default: Box download is resuming from prior download progress
Download redirected to host: cloud-images.ubuntu.com
    default:
==> default: Successfully added box 'ubuntu/focal64' (v20210826.0.0) for 'virtualbox'!
==> default: Importing base box 'ubuntu/focal64'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: VM_default_1630115273639_95765
Vagrant is currently configured to create VirtualBox synced folders with
the `SharedFoldersEnableSymlinksCreate` option enabled. If the Vagrant
guest is not trusted, you may want to disable this option. For more
information on this option, please refer to the VirtualBox manual:

  https://www.virtualbox.org/manual/ch04.html#sharedfolders

This option can be disabled globally with an environment variable:

  VAGRANT_DISABLE_VBOXSYMLINKCREATE=1

or on a per folder basis within the Vagrantfile:

  config.vm.synced_folder '/host/path', '/guest/path', SharedFoldersEnableSymlinksCreate: false
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default:
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default:
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Setting hostname...
==> default: Mounting shared folders...
    default: /vagrant => D:/Files/Omnia/595/VM
Vagrant was unable to mount VirtualBox shared folders. This is usually
because the filesystem "vboxsf" is not available. This filesystem is
made available via the VirtualBox Guest Additions and kernel module.
Please verify that these guest additions are properly installed in the
guest. This is not a bug in Vagrant and is usually caused by a faulty
Vagrant box. For context, the command attempted was:

mount -t vboxsf -o uid=1000,gid=1000,_netdev vagrant /vagrant

The error output from the command was:

: Invalid argument

// setting up ssh key here...

// works in home directory
vagrant@cit595Dev:~$ git clone git@github.com:CIS548/21fa-cit595-<student>.git
The authenticity of host 'github.com (140.82.113.3)' can't be established.
RSA key fingerprint is SHA256:nThbg6kXUpJWGl7E1IGOCspRomTxdCARLviKw6E5SY8.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'github.com,140.82.113.3' (RSA) to the list of known hosts.
warning: You appear to have cloned an empty repository.

// fails in vagrant directory
vagrant@cit595Dev:~$ cd ../../vagrant
vagrant@cit595Dev:/vagrant$ git clone git@github.com:CIS548/21fa-cit595-<student>.git
fatal: could not create work tree dir '21fa-cit595-patrickpclee1': Permission denied

**Comment by angelapwen** [comment]
Looks like this part is the error indicating that the shared folder is unable to be mounted, right?

```default: /vagrant => D:/Files/Omnia/595/VM
Vagrant was unable to mount VirtualBox shared folders. This is usually
because the filesystem "vboxsf" is not available. This filesystem is
made available via the VirtualBox Guest Additions and kernel module.
Please verify that these guest additions are properly installed in the
guest. This is not a bug in Vagrant and is usually caused by a faulty
Vagrant box. For context, the command attempted was:

mount -t vboxsf -o uid=1000,gid=1000,_netdev vagrant /vagrant

The error output from the command was:

: Invalid argument
```

**Comment by angelapwen** [comment]
I found several other users who encountered this error recently:

https://stackoverflow.com/questions/68283899/vagrant-was-unable-to-mount-virtualbox-shared-folders-error-invalid-argument

https://githubmemory.com/repo/dotless-de/vagrant-vbguest/issues/414

The answer from Stack Overflow includes a convoluted series of commands in order to get past the error; and the latter uses a VirtualBox `6.1.10` I assume the student is running `6.1.26` — @alesparza and @rstanton11 could you confirm the versions of VirtualBox that you were using when you replicated the error?

If we can confirm that the issue is resolved by using a prior release (and that this is a bug), I'm inclined to let students know to use the older version until the latest one is patched.

**Comment by kuang-da** [comment]
I just tried on my machine and I was not able to reproduce this error.

The versions of apps are up-to-date in my OS as follows:
- Windows 10, 21H1
- Vagrant 2.2.18
- VirtualBox 6.1.26

What are the versions of your apps? I will suggest you update the apps and give them another try.

Also, maybe installing [the vagrant-vbguest plugin](https://github.com/dotless-de/vagrant-vbguest) could help. But I tend to keep vagrant lightweight by not installing extra plugins if that is possible.

```
$ vagrant plugin install vagrant-vbguest
```

**Comment by kuang-da** [comment]
Interestingly, I just reproduced the error by `vagrant halt` and `vagrant up`. I will play around with it.

**Comment by kuang-da** [comment]
vagrant-vbguest partially solves the problem but it is not stable and leads to some memory issues...

```
PS D:\VMs\cis595-21fall> vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Clearing any previously set forwarded ports...
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
There was an error while executing `VBoxManage`, a CLI used by Vagrant
for controlling VirtualBox. The command and stderr is shown below.

Command: ["startvm", "3741a4b1-76ed-4f63-bc76-5e489821cdec", "--type", "headless"]

Stderr: VBoxManage.exe: error: The VM session was closed before any attempt to power it on
VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component SessionMachine, interface ISession
```

**Comment by kuang-da** [comment]
It seems to work on my side now. I made a new commit to the master.

Based on [Vagrant documentation](https://www.vagrantup.com/docs/providers/virtualbox/boxes#virtualbox-guest-additions), now they suggest installing Virtualbox Guest Additions even though they explicitly did not recommend it about a few months ago.

So the following two changes are made.

1. Based on [this reply](https://github.com/dotless-de/vagrant-vbguest/issues/414#issuecomment-894879346), I added one line of code to install some utils in Ubuntu.
2. Based on [Vagrant Manual](https://www.vagrantup.com/docs/vagrantfile/vagrant_settings#config-vagrant-plugins), I added another line of code to make sure the VM using the vbguest plugin.

To update the VM and fix the shared-folder issue, 

1. Install vagrant-vbguest by running (I restarted my OS just in case)

```
vagrant plugin install vagrant-vbguest
```

2. Go to the directory with the updated vagrant file
3. `vagrant destroy` the old instance if there is any.
4. `vagrant up` to build the new image.
5. `vagrant ssh`.
6. To ensure it works, call `vagrant halt` to turn off the VM and then `vagrant up`. Hopefully, it works now.

**Comment by alesparza** [comment]
> What are the versions of your apps? I will suggest you update the apps and give them another try.

Looks like I'm running
- Windows 10, 20H2
- Virtualbox 6.1.26
- Vagrant 2.2.18

I did a fresh install of Virtualbox and Vagrant since this is a new PC since last autumn (I ended up using a Raspberry Pi 4 in this class because this PC was in repair that semester and my backup was really old and slow)

**Comment by alesparza** [comment]
I had success with updating the vagrantfile, `vagrant destroy`, and `vagrant up` without installing the vagrant-vbguest plugin.  So just re-provisioning with the updated vagrantfile seems sufficient.

**Comment by alesparza** [comment]
Do we need to push the updated vagrantfile to Coursera and post a Note on Piazza?

**Comment by kuang-da** [comment]
> > What are the versions of your apps? I will suggest you update the apps and give them another try.
> 
> Looks like I'm running
> 
> * Windows 10, 20H2
> * Virtualbox 6.1.26
> * Vagrant 2.2.18
> 
> I did a fresh install of Virtualbox and Vagrant since this is a new PC since last autumn (I ended up using a Raspberry Pi 4 in this class because this PC was in repair that semester and my backup was really old and slow)

This is an interesting case. I love to see you achieve all the projects on Raspberry Pi 4. Another evidence showing that our projects are architecture agnostic!

**Comment by kuang-da** [comment]
> Do we need to push the updated vagrant file to Coursera and post a Note on Piazza?

Thank you for verifying the updates! That is super helpful. I am thinking about close #1 and #2 then push the changes to Coursera so that students can enjoy the shared folder and a better compiler. But that depends on whether we could close #2 in time.

**Comment by alesparza** [comment]
I could try spinning up a new VM with #2 's replacement vagrantfile and check if my old code still works with clang-10

**Comment by alesparza** [comment]
There is an open pull request in for #2 .  See comments in that issue for details

**Comment by kuang-da** [comment]
Close the issue since we have merged the PR for #2 and I have updated the access privilege.
